### PR TITLE
CORTX-34143: k8 support bundle script fails to collect cortx-rgw files.

### PR DESCRIPTION
**Pull request by shriya-deshmukh**
Signed-off-by: Shriya Deshmukh <shriya.deshmukh@seagate.com>

# Problem Statement
- If the container contains multiple compressed log '.gz' files in that case RGW support bundle collection is failing with `-9` error code, which means the process is killed via SIGKILL. It's probably due to the machine/node running out of memory or the script needing too much memory and the OS kernel was triggering the OOM killer.
- It is failing while reading the larger size `.gz` file contents using gzip.readline().

# Design
-  Instead of reading the `.gz` file contents, include these files in the support bundle if file creation time is in between the given time duration.

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: 
-  Confirm All CODACY errors are resolved [Y/N]: 

# Testing
- Are test cases updated along with code changes due to Enhancements/Bugs [Y/N]:
- Confirm that new test cases are added to regression and sanity plan files and relevant feature plan files [Y/N]:
- Confirm that Test Cases are added for new features added [Y/N]: 
- Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: 
- Confirm Testing was performed with installed RPM/K8s deployment [Y/N]:  Y

# Review Checklist 
####  Before posting the PR please ensure:
- [x] PR is self reviewed
- [ ] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [ ] New package/s added to setup.py?
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [ ] Check Acceptance Criterion is defined. 
- [ ] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified.

# KVstore/Confstore feature (changes/fixes) checklist
#### Confirm changes are made for:
- [ ] ConfStore
- [ ] KVStore
- [ ] ConfCli
- [ ] Test cases added for all above 3
- [ ] changes done in all the KVpayloads (ConsulKvPayload, IniKvPayload, KvPayload)

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

**Review by sachitanands — APPROVED** [review]
looks good

File: py-utils/src/utils/support_framework/log_filters.py [hunk 1]
```diff
@@ -201,70 +205,57 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
+                with open(in_file, 'r') as fd_in, open(op_file, 'a') as fd_out:
+                    line = fd_in.readline()
+                    while (line):
+                        regex_res = re.search(log_timestamp_regex, line)
+                        log_duration = regex_res.group(0) if regex_res else None
+                        if log_duration:
+                            # convert log timestamp to datetime object
+                            try:
+                                log_time = datetime.strptime(log_duration, datetime_format)
+                                if start_time <= log_time and log_time <= end_time:
```
Comment (cdeshmukh) [inline py-utils/src/utils/support_framework/log_filters.py:221]: Checking per line whether it falls under required timestamp is a very heavy approach i feel now.
Lets say for 10K lines in a file, worst case we will execute this for all 10k lines, O(N) time & space complexity

What if we take following approach.
1. start at the middle line in file, if start & end falls before the timestamp of this line, consider lines with timestamp before this line or vice vera. 
2. Now keep repeating this division, until either start or stop timestamp doesnt fall in the range. If start is not in the range, increase the upper bound by 100 and same way if stop timestamp is not in range, increase its range by 100 and then repeat from step 1 again.
3. In best case, we can get the loglines in log(N) complexity

PLease think over & discuss then, if you feel more complex scenarios needs to be handled. and change above algo as you feel appropriate.

and modularize this code enough, so any implementation part we can revisit, just to touch granular part

File: py-utils/src/utils/support_framework/log_filters.py [hunk 2]
```diff
@@ -201,70 +205,57 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
             if file.startswith(file_name_reg_ex):
                 in_file = os.path.join(src_dir, file)
                 if file.endswith('.gz'):
```
[the remark this answers is not in the document]
Comment (cdeshmukh) [reply py-utils/src/utils/support_framework/log_filters.py]: mod time of latest log file would still be latest among other files, which would be rotated and have old mod time anyways, so we are fine

File: py-utils/src/utils/support_framework/log_filters.py [hunk 3]
```diff
@@ -192,6 +194,8 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
             key=lambda f: os.path.getmtime(os.path.join(src_dir, f)),
             reverse=True)
         for file in list_of_files:
+            log_scope_exceeded = False
+            is_log_written_to_file = False
```
[the remark this answers is not in the document]
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py]: Done.

File: py-utils/src/utils/support_framework/log_filters.py [hunk 4]
```diff
@@ -178,12 +177,15 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
             '[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}'
         """
-        supported_file_types = ['.log', '.gz']
         invalid_chars = re.findall("[^PTDHMS0-9-:]+", duration)
         if invalid_chars:
             raise BundleError(errno.EINVAL, "Invalid duration passed! "
             + f"unexpected characters: {invalid_chars}")
 
+        include_lines_without_timestamp = False
+        supported_file_types = ['.log', '.gz']
```
[the remark this answers is not in the document]
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py]: Updated supported_file_type to = [.log, .txt]
If the file ends with '.log or .txt' then we are processing the file line by line and adding the logs to SB if it is present in the given time duration, and for other file formats we are checking the modification time of the file.

File: py-utils/src/utils/support_framework/log_filters.py (hunk 1 above)
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:219]: As discussed for now I have kept the existing approach.
Added TODO here and pushed optimized test code(not fully implemented) to the fork branch for future reference- https://github.com/Seagate/cortx-utils/compare/main...shriya-deshmukh:cortx-utils:time_filter_optimize_approach?expand=1

Comment (cdeshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:219]: >  fork branch for future reference

for future reference, & implement the optimized approach
Also above PR link opens in edit PR mode, make sure its a draft PR atleast

File: py-utils/src/utils/support_framework/log_filters.py [hunk 5]
```diff
@@ -178,93 +177,83 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
-            regex_res = re.search(timestamp_regex, line)
-            log_duration = regex_res.group(0) if regex_res else None
-            if log_duration:
-                # convert log timestamp to datetime object
+                file_extension = pathlib.Path(file).suffix
+                if file_extension not in supported_file_types:
+                    FilterLog._collect_rotated_log_file(
+                        in_file, dest_dir, start_time, end_time)
+                    continue
+                # TODO: Instead of processing the file line by line,
```
Comment (cdeshmukh) [inline py-utils/src/utils/support_framework/log_filters.py:207]: \# TODO: Parse log lines using timestamp in more optimal way, like using divide & conquer approach, to decide log lines falling with in time range

Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py]: Done

File: py-utils/src/utils/support_framework/log_filters.py (hunk 1 above)
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:219]: Updated the PR link in above comment.

**Pull request merged**